[SWEP2-11] Add users crud and roles

Hir0Yat0 opened the pull request:
how to use page access role limit right now: 
 - when making a new controller, if you want to limit access to the whole controller / all actions, add "before_action :require\_<role_name>" (the actual methods are defined in application controllers)
 - if you want to limit access to just some actions (routed methods) to a specific role and not one role to the whole controller, add "require\_<role_name>" to the first line of the method like so: 
def method_name:
  require\_<role_name>
  \<the rest of the method\>

Hir0Yat0 committed:
feat(router): added users :new

Hir0Yat0 committed:
feat: added pagy to users

Hir0Yat0 committed:
feat(user): uses same format for view as students

Hir0Yat0 committed:
refactor: fix linting

Hir0Yat0 committed:
chore: run rubocop linter

Hir0Yat0 committed:
feat(users): show roles and uuid in index and view

Hir0Yat0 committed:
feat(user): added show, new, and create handles to user

Hir0Yat0 committed:
feat(user): set default uuid before validation if uuid is blank instead of just on create

Hir0Yat0 committed:
feat(user): added  role dropdown to edit forms

Hir0Yat0 committed:
feat(user): allow only to edit users roles

Hir0Yat0 committed:
feat: added flash to shared app erb

Hir0Yat0 committed:
feat: added users roles checks methods

Hir0Yat0 committed:
refactor(user): uses defined helper method to check roles

Hir0Yat0 committed:
feat: added authorized roles for qrcodes page access

Hir0Yat0 committed:
refactor: move call render notif/notice/alert in app erb

Hir0Yat0 committed:
feat: only show accessible by roles pages links in tabs

Hir0Yat0 committed:
feat: doesnt show flash role for users

Hir0Yat0 committed:
feat(user): added organization owners roles

Hir0Yat0 committed:
feat(roles): added othe roles checks to helper functions as well

Hir0Yat0 committed:
feat: added link to login page in home drawers tab

Hir0Yat0 committed:
chore: lint _drawer erb codes

Hir0Yat0 committed:
feat(roles): attendances pages and functionalities required teachers roles

Hir0Yat0 committed:
feat(roles): student controllers require teacher roles

Hir0Yat0 committed:
feat: added methods to check require at least  one of multiple roles

Hir0Yat0 committed:
feat(roles): link to attendances and students page in home drawers nav tabs are limited to teachers now

Hir0Yat0 committed:
chore: added brakeman ignore lists

Hir0Yat0 committed:
test: updates current test setup to works with roles

Hir0Yat0 committed:
test: modify system tests to login with roles

Hir0Yat0 committed:
test: added some basic users tests

Hir0Yat0 committed:
test: added some basic user tsets

Hir0Yat0 committed:
test(user): add basic test againts accessing user info page

Hir0Yat0 committed:
test(user): check non empty on model create

Hir0Yat0 committed:
Merge branch 'main' into feature/SWEP2-11-users-crud-and-roles-v2

Hir0Yat0 committed:
chore: run linter on attedances controller

Hir0Yat0 committed:
test(user): add more role access guard tests

Hir0Yat0 committed:
chore: run linter on users controllers tests

Hir0Yat0 committed:
refactor: user tests roles loops over multiple roles

Hir0Yat0 committed:
tests: add roles tests for attendances controllers

Hir0Yat0 committed:
test: added basic roles testings to students controllers

Hir0Yat0 committed:
chore: run linter on qrcodes_controller_test.rb

Hir0Yat0 committed:
test: add basic roles tests to home controller tests

birchtrees20 commented on app/controllers/application_controller.rb:
```diff
@@ -6,4 +6,68 @@ def current_user
+      flash[:alert] = "You must be a student to access requested page."
+      redirect_to root_path
+    end
+  end
+
+  def require_unassigned
+    unless unassigned?
+      flash[:alert] = "You must be an unassigned to access requested page."
+      redirect_to root_path
+    end
+  end
+
+  def require_roles_atleast_one(roles_names_array)
+    unless authenticated? && roles_names_array.include?(current_user&.role)
+      flash[:alert] = "You do not have the necessary roles to access requested page."
+      redirect_to root_path
+    end
+  end
+
+  helper_method :admin?, :owner?, :teacher?, :student?, :unassigned?
+
+  def admin?
+    current_user&.role == "admin"
+  end
+
+  def owner?
+    current_user&.role == "owner"
+  end
+
+  def teacher?
+    current_user&.role == "teacher"
+  end
+
+  def student?
+    current_user&.role == "student"
+  end
+
+  def unassigned?
+    current_user&.role == "unassigned"
+  end
```
The logic is good but there is a lot of code repetition. Write this more compact and dynamically to reduce code redudancy.

Hir0Yat0 replied:
hmm, i think its ok here, we know  that we probably wont be increasing more roles right now, so just writing some explicit helper functions like this should be fine i think, if we need to change this then we can refactor that later

birchtrees20 replied:
I don't think refactoring this would take too much effort and it would significantly improve code readability and modularity.

this is what AI suggests at a starting point:

# Generic role requirement method
def require_role(role)
  unless current_user&.role == role
    flash[:alert] = "You must be a #{role} to access the requested page."
    redirect_to root_path
  end
end

# Specific role helper methods
%w[admin owner teacher student unassigned].each do |role|
  define_method("require_#{role}") { require_role(role) }
  define_method("#{role}?") { current_user&.role == role }
end

helper_method :admin?, :owner?, :teacher?, :student?, :unassigned?

Hir0Yat0 replied:
oh cool, guess that works in rails lol, will do

Hir0Yat0 replied:
done

birchtrees20 commented on app/controllers/users_controller.rb:
```diff
@@ -1,12 +1,45 @@
+      @pagy, @users = pagy(User.all)
+    end
+
+    # GET /users/1 or /users/1.json
+    def show
+      @user = User.all.find(params[:id]) # Fetch the user by ID from the database
+      respond_to do |format|
+        format.html { render "show" }
+        format.js
+        format.json { render json: @user.slice(:id, :created_at, :updated_at) }
+      end
+    end
```
You should implement error handling for getting nonexistent users. Currently if you try to get a user that doesn't exist your code crashes.

Hir0Yat0 replied:
will do o7

Hir0Yat0 replied:
wait, i think that only happens when we run in local host,

when i try to reproduce the same errors with students on the server rails route me properly to a 404 page, but when I try to do the same thing on localhost it shows the debug messages

so i guess its not that it crashes, but when access non-existing routes on localhosts (development environment) they show debugging messages instead

birchtrees20 commented on app/models/user.rb:
```diff
@@ -22,24 +22,24 @@ class User < ApplicationRecord
-  def generate_uuid
-    self.uuid ||= SecureRandom.uuid
+  def set_default_uuid
+    self.uuid = SecureRandom.uuid if uuid.blank?
```
why would uuid be blank?

Hir0Yat0 replied:
oh i should probably move this to a migration, its here's to for so that some users in the older versions without uuid wont crash the server.

will move this to run as  a migrations instead o7

Hir0Yat0 replied:
oh, its if blank because validation is run everytime we save / update a user not just new, so if we update only if blank, then it wont regenerate uuid for that user

birchtrees20 commented on app/views/layouts/_drawer.html.erb:
```diff
@@ -7,30 +7,68 @@
+        </li>
+      <% end %>
+
+      <% if admin? %>
+        <li>
+          <%= link_to users_path do %>
+            <i class="fa-solid fa-users"></i> Users
+          <% end %>
+        </li>        
+      <% end %>
+
+      <% if student? %>
+        <li>
+          <%= link_to "/qrcodes/" do %>
+            <i class="fas fa-qrcode"></i> My QR code
+          <% end %>
+        </li>
+      <% end %>
+
+      <% if teacher? %>
+        <li>
+          <%= link_to "/scan_qr" do %>
+            <i class="fa-solid fa-magnifying-glass"></i> QR Scan
+          <% end %>
+        </li>
+      <% end %>
+
+      <% if authenticated? %>
+        <li>
+          <%= link_to session_path, data: {turbo_method: :delete} do %>
+            <i class="fas fa-sign-out-alt"></i> Logout
+          <% end %>
+        </li>
+      <% else %>
+        <li>
+          <%= link_to new_session_path do %>
+            <i class="fas fa-sign-in-alt"></i> Login
+          <% end %>
+        </li>
+      <% end %>
```
OK

birchtrees20 commented on app/views/users/edit_1.html.erb:
```diff
@@ -0,0 +1,26 @@
+<% content_for :title, "Edit User Role" %>
+<pre>
+Form action: <%= user_path(@user) %>
+HTTP method: POST with _method=PATCH
+</pre>
+<div class="flex justify-center items-center min-h-[calc(100vh-64px)]">
+  <div class="max-w-md w-full bg-white p-6 rounded-lg shadow-md">
+    <h2 class="text-2xl font-semibold mb-4">Edit Role for <%= @user.email_address %></h2>
+
+    <%= form_with(model: @user, local: true, method: :patch, class: "space-y-4") do |form| %>
+        <div class="field mb-4">
+          <%= form.label :email_address, class: "block text-sm font-medium text-gray-700 mb-1" %>
+          <%= form.text_field :email_address, disabled: true, class: "w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm bg-gray-100" %>
+        </div>
+      
+        <div class="field mb-4">
+          <%= form.label :role, class: "block text-sm font-medium text-gray-700 mb-1" %>
+          <%= form.select :role, User::ROLES, { selected: @user.role }, { class: "w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm" } %>
+        </div>
+      
+        <div class="actions">
+          <%= form.submit "Save Changes", class: "w-full py-2 px-4 border border-transparent rounded-md shadow-sm text-sm font-medium text-white bg-indigo-600 hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500" %>
+        </div>
+    <% end %>
+  </div>
+</div>
```
What is this for and why is the naming edit and edit_1. If you need both please rename to make more clear.

Hir0Yat0 replied:
oh oops, this is the old version and forgot to delete, will delete them now o7

birchtrees20 commented on app/views/users/index.html.erb:
```diff
@@ -1,25 +1,47 @@
+      <tr>
+        <th></th>
+        <th>ID</th>
+        <th>Email</th>
+        <th>Role</th>
+        <th>UUID</th>
+        <th>Actions</th>
       </tr>
     </thead>
     <tbody>
       <% @users.each do |user| %>
-        <tr class="border-b">
-          <td class="px-4 py-2"><%= user.email_address %></td>
-          <td class="px-4 py-2"><%= user.role %></td>
-          <td class="px-4 py-2">
-            <%= link_to 'Edit Role', edit_user_path(user), class: "text-blue-500 hover:text-blue-700" %>
+        <tr>
+          <td></td>
+          <td><%= user.id%></td>
+          <td><%= user.email_address%></td>
+          <td><%= user.role%></td>
+          <td><%= user.uuid%></td>
+          <td>
+            <div class="flex space-x-2">
+              <%= link_to user, class: "hover:text-blue-500 text-blue-700" do %>
+                <i class="fas fa-eye" title="Show"></i> View
+              <% end %>
+              <%= link_to edit_user_path(user), class: "hover:text-yellow-500 text-yellow-700" do %>
+                <i class="fas fa-edit" title="Edit"></i> Edit
+              <% end %>
+            </div>
           </td>
         </tr>
       <% end %>
     </tbody>
   </table>
 </div>
+<div class="flex justify-center items-center">
+  <%== pagy_nav(@pagy) %>
+</div>
```
Unless this is best practice and I'm wrong, do you think renaming this to something like user index would be more effective in signaling that this is the user list?

Hir0Yat0 replied:
renaming what? the file name?

birchtrees20 replied:
Yea I was talking about the file name index.html.erb. I was just wondering if the naming is clear enough but if it takes too much change and it isn't worth it then thats fine

Hir0Yat0 replied:
oh this one is by convention from rails, it will auto routes requests for that /\<controllers\> to index, and is differentiate from other controllers by the directory  name

birchtrees20 commented on test/models/user_test.rb:
```diff
@@ -52,4 +52,9 @@ class UserTest < ActiveSupport::TestCase
     user = User.new email_address: "a222@bbb.com", password: "password"
     assert_not user.save
   end
+
+  test "should not have empty uuid on create" do
+    user = User.new email_address: "anewuser@abc.com", password: "5555555555", uuid: SecureRandom.uuid
+    assert user.uuid.present?
+  end
```
why was there an issue with empty uuid in the first place when every new user gets generated an id anyway?

Hir0Yat0 replied:
no issue here, this test is to ensure that every new users indeed do get their uuid generated

birchtrees20 replied:
ok

birchtrees20 reviewed (changes requested):
- very good and extensive tests
- some minor code refactoring / renaming to consider
- add error-handling and tests for getting nonexistent users

Overall very solid pull request and only some small changes needed.

Please look at my comments and make changes. Feel free to contest me on my comments if you disagree.

Hir0Yat0 committed:
fix: remove old version of user view's edit and index page

Hir0Yat0 committed:
refactor: define roles checking methods with rails metaprogramming

Hir0Yat0 committed:
feat: ran migration called "ChangeStudentIdToBeNullable"

i ran a migration called "ChangeStudentIdToBeNullable" that was somewhen and somehow  added to here without me noticing and this commit is the result of that migration being run

Hir0Yat0 committed:
feat: add migration to fill users with empty uuid with uuid

Hir0Yat0 committed:
fix: swap around login and logout guard in dashboard drawer

Hir0Yat0 committed:
feat: run migration to make user uuid non nullable

birchtrees20 reviewed (approved):
ok
